django/db/models/query.py

```diff
@@ -689,7 +689,16 @@ def in_bulk(self, id_list=None, *, field_name='pk'):
         """
         assert not self.query.is_sliced, \
             "Cannot use 'limit' or 'offset' with in_bulk"
-        if field_name != 'pk' and not self.model._meta.get_field(field_name).unique:
+        unique_constraints = [
+            constraint.fields[0]
+            for constraint in self.model._meta.total_unique_constraints
+            if len(constraint.fields) == 1
```

Comment:
I think we also need to exclude [conditional constraints](https://docs.djangoproject.com/en/3.0/ref/models/constraints/#condition) here.

Reply:
That's already handled by `total_unique_constraints`:
https://github.com/django/django/blob/ad811335bdd57d65dddbc05375ec3a70b59c96fb/django/db/models/options.py#L832-L841

Reply:
Ah! Maybe include a test for the scenario then.